robotPositioner/robotPositioner.h

```diff
@@ -0,0 +1,382 @@
+
+            // orientation of Target with respect to the line of sight from the observer to the target
+            double theta = m_heading + m_bearingFromGoal - m_goalAngle; 
+            theta = angleWrapAround(theta);
+
+            // curveness value: k
+            double k = -(1/m_distanceFromGoal)* ( m_k2* (m_bearingFromGoal- atan(-m_k1*theta*PI/180.0)*180.0/PI) + 
+                1+(m_k1/(1+pow(m_k1*theta,2)))*sin(m_bearingFromGoal*PI/180.0)*180.0/PI);
+
+            // current velocity
+            double v = m_max_velocity/(1+m_beta*pow(abs(k),m_alpha));
+
+            // formula for curvy path
+            double w = k*v;
+       
+            // if we are close to the goal and the heading error is less than the allowed, stop the robot
+            if (m_distanceFromGoal < m_stopping_distance && closeTo(m_bearingFromGoal,m_allowed_heading_error)) {
+
+                stopFlag = false;
+                m_bot->tank(0,0);
+            } else {
+                driveRobotWithVelocities(v,w, m_robot_r, m_robot_D);
+            }
+
+            std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        }
+    }
+
+    // drives the robot wheels using v and w as velocities which get translated to 
+    // indiviual wheel velocities. Conversion from these velocities to individual 
+    // wheel velocities is done via solving simultaneous equations
+    // parameters:
+    //             v      : translational velocity
+    //             w      : rotational velocity
+    //             robot_r: wheel radius
+    //             robot_d: axis length
+    template <class T, class S>
+    void driveRobotWithVelocities(const T v, const T w, const S robot_r, const S robot_d) {
+        // v = r * (Vl+Vr)/2
+        // w = r * (Vr-Vl)/D
+        T a =  robot_r/2;
+        T b =  robot_r/2;
+        T c =  robot_r / robot_d;
+        T d = -(robot_r / robot_d);
+
+        T e=v;
+        T f=w;
+        T Vl,Vr;
+
+        // determinant
+        T det = a*d - b*c;
+
+        // if the determinant is not zero, there is a solution
+        if (det != 0) {
+            Vl = (e*d - b*f)/det;
+            Vr = (a*f - e*c)/det;
+        }
+        // drive robot 
+        m_bot->tank(Vl, Vr);
+    }
+
+    // captures a frame
+    void captureCameraFrame() {
+        const cv::Size unwrapSize(180,50);
+        cv::Mat output(m_cam->getSuperPixelSize(), CV_8UC1);
+        cv::Mat unwrapped(unwrapSize, CV_8UC1);
+
+        if (m_cam->captureSuperPixelGreyscale(output)) {
+            m_unwrapper.unwrap(output, unwrapped);
+            char fileName[255];
+            sprintf(fileName, "image_%u.png", m_imageNumber++);
+            cv::imwrite(fileName, unwrapped);
+            std::cout << " image " << m_imageNumber << " captured " << std::endl;
+        }
+    }
+
+//-----------------PUBLIC API---------------------------------------------------------------------
+public:
+
+    // constructor
+    robotPositioner(
```

Comment:
I think it would be much nicer if this class just took references to a ``BoBRobotics::Robots::Tank`` and a ``Vicon::UDPClient<ObjectData>`` - this would:
- Reduce the number of parameters to the constructor
- Make it more flexible
- Make it easier to integrate into a larger program

Reply:
Agreed. I think the ``Tank`` object will have to be a pointer though rather than a reference because it's an abstract class and C++ is funny about that

Reply:
I **believe** references to abstract classes **should** work:

```c++
#include <iostream>

class Bar
{
public:
    virtual void update() = 0;
};

class Foo : public Bar
{
public:
    virtual void update() override
    {
        std::cout << "Updating foo" << std::endl;
    }
};

void updateBar(Bar &bar)
{
    bar.update();
}

int main()
{
    Foo a;
    updateBar(a);
    
    return EXIT_SUCCESS;
}
```